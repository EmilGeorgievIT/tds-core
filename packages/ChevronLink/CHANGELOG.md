# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.0.3](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@2.0.2...@tds/core-chevron-link@2.0.3) (2019-04-22)

**Note:** Version bump only for package @tds/core-chevron-link





## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@2.0.1...@tds/core-chevron-link@2.0.2) (2019-04-16)

**Note:** Version bump only for package @tds/core-chevron-link





## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@2.0.0...@tds/core-chevron-link@2.0.1) (2019-04-11)

**Note:** Version bump only for package @tds/core-chevron-link





# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.14...@tds/core-chevron-link@2.0.0) (2019-04-08)


### Features

* **core-chevron-link:** upgrade to styled-components ([54750bf](https://github.com/telusdigital/tds/commit/54750bf))


### BREAKING CHANGES

* **core-chevron-link:** upgrade to styled-components





## [1.1.14](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.13...@tds/core-chevron-link@1.1.14) (2019-04-04)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.13](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.12...@tds/core-chevron-link@1.1.13) (2019-03-28)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.12](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.11...@tds/core-chevron-link@1.1.12) (2019-03-28)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.11](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.10...@tds/core-chevron-link@1.1.11) (2019-03-20)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.10](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.9...@tds/core-chevron-link@1.1.10) (2019-03-20)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.9](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.8...@tds/core-chevron-link@1.1.9) (2019-03-18)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.8](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.7...@tds/core-chevron-link@1.1.8) (2019-03-18)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.7](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.6...@tds/core-chevron-link@1.1.7) (2019-03-15)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.6](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.5...@tds/core-chevron-link@1.1.6) (2019-03-11)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.5](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.4...@tds/core-chevron-link@1.1.5) (2019-03-05)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.4](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.3...@tds/core-chevron-link@1.1.4) (2019-02-06)

**Note:** Version bump only for package @tds/core-chevron-link





## [1.1.3](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.2...@tds/core-chevron-link@1.1.3) (2019-01-10)

### Bug Fixes

- **core-chevron-link:** replace airbnb proptypes with util-prop-types ([d7b6ab7](https://github.com/telusdigital/tds/commit/d7b6ab7))

<a name="1.1.2"></a>

## [1.1.2](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.1...@tds/core-chevron-link@1.1.2) (2018-12-19)

### Bug Fixes

- **core-chevron-link:** use shared typography stylesheet ([1a71b0e](https://github.com/telusdigital/tds/commit/1a71b0e))

<a name="1.1.1"></a>

## [1.1.1](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.1.0...@tds/core-chevron-link@1.1.1) (2018-11-13)

### Bug Fixes

- **core-chevron-link:** ie11 overlfow issue ([66f3204](https://github.com/telusdigital/tds/commit/66f3204)), closes [#642](https://github.com/telusdigital/tds/issues/642)

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.0.2...@tds/core-chevron-link@1.1.0) (2018-07-17)

### Features

- **core-chevron-link:** add A11yContent props ([7ac29fd](https://github.com/telusdigital/tds/commit/7ac29fd))
- **core-chevron-link:** support A11yContent as a child component ([b0bb956](https://github.com/telusdigital/tds/commit/b0bb956))

<a name="1.0.2"></a>

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.0.1...@tds/core-chevron-link@1.0.2) (2018-04-19)

### Bug Fixes

- **core-chevron-link:** changed tag from div to span ([76429dd](https://github.com/telusdigital/tds/commit/76429dd))

<a name="1.0.1"></a>

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-chevron-link@1.0.0...@tds/core-chevron-link@1.0.1) (2018-04-02)

### Bug Fixes

- **core-box:** update tests ([c0f0343](https://github.com/telusdigital/tds/commit/c0f0343))
