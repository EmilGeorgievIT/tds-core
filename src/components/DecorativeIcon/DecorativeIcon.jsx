import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../safeRest'
import joinClassNames from '../../utils/joinClassNames'
import capitalize from '../../capitalize'

import styles from './DecorativeIcon.modules.scss'

const DecorativeIcon = ({ symbol, variant, size, ...rest }) => {

  const classes = joinClassNames(
    styles[`icon${capitalize(symbol)}`],
    variant && styles[variant],
    size && styles[`size${size}`]
  )

  return (
    <i {...safeRest(rest)} className={classes} aria-hidden="true" />
  )
}

DecorativeIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'incomplete',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times'
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'inverted',
    'error'
  ]),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
}

DecorativeIcon.defaultProps = {
  variant: undefined,
  size: 24
}

export default DecorativeIcon
