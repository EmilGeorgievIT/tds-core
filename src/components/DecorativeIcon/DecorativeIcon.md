### Available Icons

```
<div className="docs--horizontal-spacing">
  <DecorativeIcon symbol="caretDown" />
  <DecorativeIcon symbol="caretUp" />
  <DecorativeIcon symbol="checkmark" />
  <DecorativeIcon symbol="leftChevron" />
  <DecorativeIcon symbol="chevron" />
  <DecorativeIcon symbol="exclamationPointCircle" />
  <DecorativeIcon symbol="expander" />
  <DecorativeIcon symbol="hamburger" />
  <DecorativeIcon symbol="incomplete" />
  <DecorativeIcon symbol="location" />
  <DecorativeIcon symbol="minus" />
  <DecorativeIcon symbol="plus" />
  <DecorativeIcon symbol="questionMarkCircle" />
  <DecorativeIcon symbol="spyglass" />
  <DecorativeIcon symbol="times" />
</div>
```
