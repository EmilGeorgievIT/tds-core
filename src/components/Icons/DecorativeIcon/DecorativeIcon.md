### Available Icons

```
<div className="docs--horizontal-spacing">
  <DecorativeIcon symbol="caretDown" />
  <DecorativeIcon symbol="caretUp" />
  <DecorativeIcon symbol="checkmark" />
  <DecorativeIcon symbol="leftChevron" />
  <DecorativeIcon symbol="chevron" />
  <DecorativeIcon symbol="exclamationPointCircle" />
  <DecorativeIcon symbol="expander" />
  <DecorativeIcon symbol="hamburger" />
  <DecorativeIcon symbol="incomplete" />
  <DecorativeIcon symbol="location" />
  <DecorativeIcon symbol="minus" />
  <DecorativeIcon symbol="plus" />
  <DecorativeIcon symbol="questionMarkCircle" />
  <DecorativeIcon symbol="spyglass" />
  <DecorativeIcon symbol="times" />
</div>
```

### Modifying color

Use the `variant` prop to alter the icon's color. Each variant has semantic meaning.

```
<div className="docs--horizontal-spacing">
  <DecorativeIcon symbol="checkmark" variant="primary" />
  <DecorativeIcon symbol="incomplete" variant="secondary" />
  <DecorativeIcon symbol="exclamationPointCircle" variant="error" />
</div>
```

### Adjusting size

Use the `size` prop to adjust the icon's size.

```
<div className="docs--horizontal-spacing">
  <DecorativeIcon symbol="chevron" variant="primary" size={16} />
  <DecorativeIcon symbol="exclamationPointCircle" variant="error" size={24} />
</div>
```
