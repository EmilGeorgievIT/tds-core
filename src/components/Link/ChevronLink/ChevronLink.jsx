import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../../safeRest'
import Icon from '../../../old-components/Icon/Icon'
import { warn } from '../../../warn'

import styles from './ChevronLink.modules.scss'

const getClassName = (variant) => {
  switch (variant) {
    case 'secondary':
      return styles.secondary
    case 'inverted':
      return styles.inverted
    default:
      return styles.primary
  }
}

const getIcon = (glyph, className) => (
  <span className={className}>
    <Icon glyph={glyph} aria-hidden="true" />
  </span>
)

/**
 * A Call to Action link.
 *
 * <span class="docs--badge green">new!</span> <span class="docs--badge purple">v0.21.0</span>
 */
const ChevronLink = ({ reactRouterLinkComponent, variant, direction, children, ...rest }) => {
  if ((reactRouterLinkComponent || rest.to) &&
      !(reactRouterLinkComponent && rest.to)) {
    warn('Chevron Link', 'The props `reactRouterLinkComponent` and `to` must be used together.')
  }

  return React.createElement(

    reactRouterLinkComponent || 'a',
    {
      ...safeRest(rest),
      className: getClassName(variant)
    },
    direction === 'left' ? getIcon('left-chevron', styles.leftChevron) : undefined,
    children,
    direction === 'right' ? getIcon('chevron', styles.rightChevron) : undefined
  )
}

ChevronLink.propTypes = {
  /**
   * The style.
   */
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'inverted'
  ]),
  /**
   * The chevron's direction and placement.
   */
  direction: PropTypes.oneOf([
    'left',
    'right'
  ]),
  /**
   * React Router Link component.
   */
  reactRouterLinkComponent: PropTypes.func,
  /**
   * Target URL (if using 'reactRouterLinkComponent')
   */
  to: PropTypes.string,
  /**
   * Target URL.
   */
  href: PropTypes.string,
  /**
   * The label.
   */
  children: PropTypes.string.isRequired
}
ChevronLink.defaultProps = {
  variant: 'primary',
  direction: 'right',
  reactRouterLinkComponent: null,
  to: null,
  href: null
}

export default ChevronLink
