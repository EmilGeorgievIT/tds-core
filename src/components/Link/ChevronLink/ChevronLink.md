```
<Link.Chevron>Hello world</Link.Chevron>
```
