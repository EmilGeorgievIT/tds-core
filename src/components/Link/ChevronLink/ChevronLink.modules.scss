@import '../../../scss/settings/colours';
@import '../../../scss/settings/typography';

.link {
  @include telus-font-family;
  @include helvetica-neue-roman-55;

  composes: medium from '../../Typography/Text/Text.modules.scss';

  &:hover {
    .rightChevron {
      transform: translateX(0.25rem);
    }
    .leftChevron {
      transform: translateX(-0.25rem);
    }
  }
}

.primary {
  composes: link;

  &:link,
  &:visited {
    color: $color-primary;
  }
}

.secondary {
  composes: link;

  &:link,
  &:visited {
    color: $color-secondary;
  }
}

.inverted {
  composes: link;

  &:link,
  &:visited {
    color: $color-white;
  }
}

.chevron {
  display: inline-block;
  transition: transform 300ms;
}

.rightChevron {
  composes: chevron;

  margin-left: 0.5rem;
}

.leftChevron {
  composes: chevron;

  margin-right: 0.5rem;
}
