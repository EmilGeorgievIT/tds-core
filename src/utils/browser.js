/* global document */
/* eslint-disable import/prefer-default-export */
// Provide a way to import the browser's document, which allows it to be mocked in tests

export const doc = document
