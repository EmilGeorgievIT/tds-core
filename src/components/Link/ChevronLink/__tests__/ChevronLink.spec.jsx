import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { warn } from '../../../../utils/warn'

import ChevronLink from '../ChevronLink'
import DecorativeIcon from '../../../Icons/DecorativeIcon/DecorativeIcon'

jest.mock('../../../../utils/warn', () => (
  { warn: jest.fn() }
))

describe('ChevronLink', () => {
  const doShallow = (overrides = {}) => shallow(
    <ChevronLink {...overrides}>Go home</ChevronLink>
  )

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const link = doShallow()

    expect(toJson(link)).toMatchSnapshot()
  })

  it('is an anchor HTML element when using the href attribute', () => {
    const link = doShallow({ href: 'http://telus.com' })

    expect(link).toHaveTagName('a')
    expect(link).toHaveProp('href', 'http://telus.com')
  })

  it('renders a react router link element when passed as a prop', () => {
    const MyLink = () => <span />
    const link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(link).toMatchSelector('MyLink')
  })

  it('must use `reactRouterLinkComponent` and `to` props together', () => {
    const MyLink = () => <span />
    let link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(warn).toHaveBeenCalled()

    jest.clearAllMocks()

    link = doShallow({ to: '/about' })

    expect(link).toHaveProp('to')
    expect(warn).toHaveBeenCalled()
  })

  it('has a chevron icon', () => {
    let link = doShallow({ href: 'https://telus.com' })
    expect(link).toContainReact(
      <span className="rightChevron">
        <DecorativeIcon symbol="chevron" size={16} />
      </span>
    )

    link = doShallow({ href: 'https://telus.com', direction: 'left' })
    expect(link).toContainReact(
      <span className="leftChevron">
        <DecorativeIcon symbol="leftChevron" size={16} />
      </span>
    )
  })

  it('can have specific variants', () => {
    let link = doShallow({ href: 'https://telus.com' })
    expect(link).toHaveClassName('primary')

    link = doShallow({ href: 'https://telus.com', variant: 'secondary' })
    expect(link).toHaveClassName('secondary')

    link = doShallow({ href: 'https://telus.com', variant: 'primary' })
    expect(link).toHaveClassName('primary')

    link = doShallow({ href: 'https://telus.com', variant: 'inverted' })
    expect(link).toHaveClassName('inverted')
  })

  it('passes additional attributes to the link element', () => {
    const link = doShallow({ id: 'the-link', role: 'button' })

    expect(link).toHaveProp('id', 'the-link')
    expect(link).toHaveProp('role', 'button')
  })

  it('does not allow custom CSS', () => {
    const link = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(link).not.toHaveProp('className', 'my-custom-class')
    expect(link).not.toHaveProp('style')
  })
})
