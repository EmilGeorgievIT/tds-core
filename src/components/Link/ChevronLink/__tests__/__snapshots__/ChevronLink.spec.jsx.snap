// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChevronLink renders 1`] = `
<a
  class="primary"
  href="test.com"
>
  Go home
  <span
    class="rightChevron"
  >
    <i
      aria-hidden="true"
      class="iconChevron size16"
    >
      
    </i>
  </span>
</a>
`;
